src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java

```diff
@@ -501,9 +501,73 @@ public synchronized boolean reserve(List<LockableResource> resources,
 		return true;
 	}
 
+	private void unreserveResources(@Nonnull List<LockableResource> resources) {
+		for (LockableResource l : resources) {
+			l.unReserve();
+		}
+		save();
+	}
 	public synchronized void unreserve(List<LockableResource> resources) {
+		// make sure there is a list of resources to unreserve
+		if (resources == null || (resources.size() == 0)) {
+			return;
+		}
+		List<String> resourceNamesToUnreserve = new ArrayList<>();
 		for (LockableResource r : resources) {
-			r.unReserve();
+			resourceNamesToUnreserve.add(r.getName());
+		}
+
+		// check if there are resources which can be unlocked (and shall not be unlocked)
+		List<LockableResource> requiredResourceForNextContext = null;
+		QueuedContextStruct nextContext = this.getNextQueuedContext(resourceNamesToUnreserve, false);
+
+		// no context is queued which can be started once these resources are free'd.
+		if (nextContext == null) {
+			unreserveResources(resources);
+			return;
+		}
+
+		// remove context from queue and process it
+		requiredResourceForNextContext = checkResourcesAvailability(nextContext.getResources(), null, resourceNamesToUnreserve);
+		this.queuedContexts.remove(nextContext);
```

Comment:
So, now both `unlock` and `unreserve` are modifying `queuedContexts`, could that lead to illegal concurrent modifications? Concerning.

Reply:
Or maybe when a resource is reserved there is no room for a call to unlock and we are safe then...

Reply:
Honestly, I hadn't even thought about that, so yeah, we should do something.

Reply:
@abayer Any updates on this PR?

Reply:
Maybe I missed something, but I achieved the same thing with a single line in PR #87. Simply calling `unlock(resources, null);` after it has been unreserved, which means the resource is processed the same way as if it was unlocked by a job. Since the `unlock` method is already synchronized there shouldn't be any concurrency issue.

Reply:
@amuniz - I think I'm working around this by using `unqueueContext(StepContext)` instead of `queuedContexts.remove(QueuedContextStruct)` in my next commit.

Reply:
I think so 👍